runtime/src/bank.rs

```diff
@@ -6428,6 +6428,11 @@ impl Bank {
             self.reconfigure_token2_native_mint();
         }
         self.ensure_no_storage_rewards_pool();
+
+        if new_feature_activations.contains(&feature_set::cap_accounts_data_len::id()) {
+            const ACCOUNTS_DATA_LEN: u64 = 50_000_000_000;
```

Comment:
What happens when we reach this limit, no new accounts can be created? How quickly is the account size currently growing on mainnet?  Do we know the number that causes a 128GB RAM validator to die?  Some kind of warning when we're getting close to this limit seems necessary, so devs can do something before the chain effectively goes down

It'd be nice to setup the code such that it's in a nice position to bump the number again in the future because we certainly will need to do so, and probably multiple times.

Also I don't love `accounts_data_len`.  That immediately means, "length of a single account" to me.  So my dumb side is wondering why we are forcing the length of an account to 50GB.  Perhaps something like  `max_total_account_data_usage`?  Lets bikeshed this name for a while to get it more precise.

Reply:
`MAX_TOTAL_ACCOUNT_DATA_SIZE` that Jack referenced [here](https://github.com/solana-labs/solana/issues/21604#issuecomment-985889427) seems nice too :)

Reply:
Ah, I think I wasn't specific enough. This change sets the _current_ size of the total accounts data, not the max. The max/limit is here:
https://github.com/solana-labs/solana/blob/6b0d34d70dc08beb3d94cc8e30423883de73646b/program-runtime/src/accounts_data_meter.rs#L6-L10

To enable the feature, there needs to be a deterministic starting/seed value, which is what this PR is doing. 

> What happens when we reach this limit, no new accounts can be created?

Correct. Banking Stage will drop transactions if it thinks the data size will exceed the limit (by checking CreateAccount-like transactions in CostTracker). Replay Stage will fail _the block_ if any _instruction_ exceeds the cap. This is because transactions processing is in parallel, and we need a cluster-wide deterministic set of transaction processing results. [This comment](https://github.com/solana-labs/solana/pull/23422#issuecomment-1055928671) has a specific example.

> It'd be nice to setup the code such that it's in a nice position to bump the number again in the future because we certainly will need to do so, and probably multiple times.

I agree. I think that changing the `MAX_ACCOUNTS_DATA_LEN` constant will be sufficient. (This reply may not be salient anymore if my first point cleared up miscommunication.)

> Some kind of warning when we're getting close to this limit seems necessary, so devs can do something before the chain effectively goes down.

Yes, I've asked the devops team to set up an alert. I need to follow up with them, as [that metric](https://metrics.solana.com:8888/sources/1/chronograf/data-explorer?query=SELECT%20mean%28%22accounts_data_len%22%29%20AS%20%22mean_accounts_data_len%22%2C%20max%28%22accounts_data_len%22%29%20AS%20%22max_accounts_data_len%22%20FROM%20%22mainnet-beta%22.%22autogen%22.%22bank-forks_set_root%22%20WHERE%20time%20%3E%20%3AdashboardTime%3A%20AND%20time%20%3C%20%3AupperDashboardTime%3A%20GROUP%20BY%20time%28%3Ainterval%3A%29%20FILL%28null%29) is now live and can be tracked.

[This dashboard](https://metrics.solana.com:8888/sources/1/dashboards/15) shows the total accounts data size along with the total number of accounts (refer to "Number of Accounts (from calculate_accounts_hash_without_index)" for more accurate numbers (i.e. without additional filler accounts)).

> Also I don't love `accounts_data_len`.

I can rename this. Note this name is also in instruction errors, transaction errors, snapshot code, AccountsDataMeter, and CostTracker; if any of that is either (1) an issue, and/or (2) should be changed also. To see all the spots quickly-ish, refer to all the PRs linked in the main issue, #21604. (This reply may not be salient anymore if my first point cleared up miscommunication.)

Reply:
oh I see.  So this the "total_accounts_data_size".  It's not really a length at all.

Thanks for the additional context.  There's no way to compute the actual value, once, at feature activation?  It's a little triggering that this would be so far off for testnet in particular, but yeah I guess it doesn't matter ultimately.  If this was a new cluster starting from block 0, we'd have the correct total_accounts_data_size right?  The Bank updates its data size on freezing?

Reply:
> oh I see. So this the "total_accounts_data_size". It's not really a length at all.

Yeah... only in the "Rust" meaning w.r.t. the "length" of a `Vec<u8>`. I saw other similar variables with `_len`, so I went with that. If I could snap my fingers, I'd switch 'em to `_size`.

> There's no way to compute the actual value, once, at feature activation? 

Kind of yes, kind of no. To get the total accounts data size, we have to basically do a `scan_accounts()` on all of AccountsDb. This takes 2-5 minutes on a GCE box beefy enough for being a validator. I figured a 2-5 minute pause for every node would be bad, esp if 100% accuracy is not useful. I'd actually much prefer a better way than set-a-random-number-that's-close-enough, so if there are other ways to calculate this number I'd be very interested.

> If this was a new cluster starting from block 0, we'd have the correct total_accounts_data_size right?

When starting from block 0, this eventually calls `Bank::default_with_accounts()`. That function _does_ perform the `scan_accounts()` to get the correct starting total accounts data size. 

(The assumption here is that the total number of accounts is very small, thus the call to `scan_accounts()` is fast. Another assumption is that block 0 can tolerate a one-time latency hit.)

Additionally, `apply_feature_activations()` is _not_ called, so the `accounts_data_len` will not be re-set to 50 B.

> The Bank updates its data size on freezing?

The Bank updates its data size (1) after processing messages (`Bank::execute_loaded_transaction()`), and (2) after collecting rent (since rent collection can cause accounts to go to zero lamports and thus get cleaned up). In normal runtime, new banks get their initial accounts data size from their parent.